returnn/hdf.py

```diff
@@ -281,3 +281,71 @@ def run(self):
             audio.close()
 
         out_hdf.close()
+
+
+class RasrAlignmentDumpHDFJob(Job):
+    """
+    This Job reads Rasr alignment caches and dump them in hdf files.
+    """
+
+    def __init__(
+        self,
+        alignment_caches: List[tk.Path],
+        allophone_file: tk.Path,
+        state_tying_file: tk.Path,
+        returnn_root: Optional[tk.Path] = None,
+    ):
+        """
+        :param list[Path]: alignment_caches
+        :param Path: allophone_file
+        :param Path: state_tying_file
+        :param Optional[Path] returnn_root: file path to the RETURNN repository root folder
+        """
+        self.alignment_caches = alignment_caches
+        self.allophone_file = allophone_file
+        self.state_tying_file = state_tying_file
+        self.out_hdf_files = [self.output_path(f"data.hdf.{d}", cached=False) for d in range(len(alignment_caches))]
+        self.returnn_root = returnn_root
+        self.rqmt = {"cpu": 1, "mem": 8, "time": 0.5}
+
+    def tasks(self):
+        yield Task("run", resume="run", rqmt=self.rqmt, args=range(1, (len(self.alignment_caches) + 1)))
+
+    def run(self, task_id):
+        from i6_core.lib.rasr_cache import FileArchive
+        import numpy as np
+
+        state_tying = dict(
+            (k, int(v)) for l in open(self.state_tying_file.get_path()) for k, v in [l.strip().split()[0:2]]
+        )
+
+        alignment_cache = FileArchive(
+            self.alignment_caches[min(task_id - 1, len(self.alignment_caches) - 1)].get_path()
+        )
+        alignment_cache.setAllophones(self.allophone_file.get_path())
+
+        returnn_root = None if self.returnn_root is None else self.returnn_root.get_path()
+        SimpleHDFWriter = get_returnn_simple_hdf_writer(returnn_root)
+        out_hdf = SimpleHDFWriter(filename=self.out_hdf_files[task_id - 1], dim=1)
+
+        for file in alignment_cache.ft:
+            info = alignment_cache.ft[file]
+            if info.name.endswith(".attribs"):
+                continue
+            seq_name = info.name
+
+            # alignment
+            targets = []
+            alignment = alignment_cache.read(file, "align")
+            alignmentStates = ["%s.%d" % (alignment_cache.allophones[t[1]], t[2]) for t in alignment]
+            for allophone in alignmentStates:
+                targets.append(state_tying[allophone])
+
+            data = np.array(targets).astype("int32")
```

Comment:
I know returnn expects int32 by default, but you can specify that the dtype is int16 as well. And saving the data as int16 will save 50% of the storage space (and will most likely be enough for the number of outputs that we usually use).

There can be cases (compatibility to old setups, or dense triphone cart) where we still need int32, so maybe we make it configurable.

I propose int16 as default purely based on consistency to https://github.com/rwth-i6/i6_core/blob/07b40104764da1ae2b3209b2fd69964e4497e977/returnn/hdf.py#L214

Reply:
but when do you use HDFs in RASR? Do we have that in a setup?

Reply:
And @michelwi, should it then here not be better uint16?

Reply:
> The problem here is not returnn but rasr. for feature scorers that rely on tensorflow node one prepares the data as int32. I personally do not want to need to redefine this detail.

Exactly this is an argument to make it configurable. 
We can discuss what should be the default type. I tend to agree with @JackTemaki on uint16.

Thinking about it: If you talk about "tensorflow node", you mean the flow node? But in rasr flow you do have generic type convert nodes, so this could still be used. Concerning the trade-off "having N more lines in your config" vs "double storage usage for the data" I would always decide in favor of storage.

Reply:
This is the case for the factored hybrid that needs to feed the context. You prepare the data in the feature scorer in rasr and you need to define the type. 
There is also a more important problem. I for example use dense tying and this does not fit in 16 bits (~(42**3)*6). This is also the reason why I still cannot integrate my triphone model in rasr search.
Please consider also this edge cases (factored hybrid) when you decide for the default. This should be more than 16 bits.

Reply:
> but when do you use HDFs in RASR? Do we have that in a setup?

See above response. If you need the label as context that is fed.

Reply:
> I for example use dense tying and this does not fit in 16 bits (~(42**3)*6). 

yes, definitively should be configurable then!